official/recommendation/ncf_keras_main.py

```diff
@@ -0,0 +1,241 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""NCF framework to train and evaluate the NeuMF model.
+
+The NeuMF model assembles both MF and MLP models under the NCF framework. Check
+`neumf_model.py` for more details about the models.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import heapq
+import json
+import logging
+import math
+import multiprocessing
+import os
+import signal
+import typing
+
+# pylint: disable=g-bad-import-order
+import numpy as np
+from absl import app as absl_app
+from absl import flags
+import tensorflow as tf
+# pylint: enable=g-bad-import-order
+
+from tensorflow.contrib.compiler import xla
+from official.datasets import movielens
+from official.recommendation import constants as rconst
+from official.recommendation import data_pipeline
+from official.recommendation import data_preprocessing
+from official.recommendation import ncf_common
+from official.recommendation import neumf_model
+from official.utils.flags import core as flags_core
+from official.utils.logs import hooks_helper
+from official.utils.logs import logger
+from official.utils.logs import mlperf_helper
+from official.utils.misc import distribution_utils
+from official.utils.misc import model_helpers
+
+
+FLAGS = flags.FLAGS
+
+
+def main(_):
+  with logger.benchmark_context(FLAGS), \
+      mlperf_helper.LOGGER(FLAGS.output_ml_perf_compliance_logging):
+    mlperf_helper.set_ncf_root(os.path.split(os.path.abspath(__file__))[0])
+    if FLAGS.tpu:
+      raise ValueError("NCF in Keras does not support TPU for now")
+    run_ncf(FLAGS)
+
+
+def run_ncf(_):
+  """Run NCF training and eval with Keras."""
+  params = ncf_common.parse_flags(FLAGS)
+
+  distribution = ncf_common.get_distribution_strategy(params)
+
+  num_users, num_items, num_train_steps, num_eval_steps, producer = (
+      ncf_common.get_inputs(params))
+
+  params["num_users"], params["num_items"] = num_users, num_items
+  producer.start()
+  model_helpers.apply_clean(flags.FLAGS)
+
+  # TODO(shiningsun): Both MirroredStrategy and OneDeviceStrategy error out.
+  # Find out why and change the distribute to distribution
+  with distribution_utils.MaybeDistributionScope(None):
+    keras_model = _get_keras_model(params)
+
+    optimizer = ncf_common.get_optimizer(params)
+    keras_model.compile(optimizer=optimizer)
+
+    train_input_dataset, eval_input_dataset = _get_train_and_eval_data(producer, params)
+
+    keras_model.fit(train_input_dataset,
+        epochs=FLAGS.train_epochs,
+        steps_per_epoch=num_train_steps,
+        callbacks=[IncrementEpochCallback(producer)],
```

Comment:
I think I discussed this with Taylor, and it cannot be easily achieved. 
The producer is putting stuff into the queue, it only maintains the left side of the sliding window. Here I am trying to move the other side of the sliding window. I don't think producer is able to do this.

Reply:
To provide some context on why this is the way it is (and how we might be able to make it cleaner):

There are two paths for getting data from the data generation thread to the training process. The GPU case is easy because it it all local, so the generation thread simply puts data onto a queue while the main thread pulls data off that queue using a generator and `tf.data.Dataset.from_generator`.  So in that case you could absolutely increment the count when the generator finished yielding before returning. (Which I think is what you're getting at.)

However in the TPU case, the data generation thread writes the data to TFRecord files, and the contents of those files is streamed to the TPU via `StreamingFilesDataset`. In that case there is no Python inside the `Dataset`, and it therefore cannot increment the counter. So in order to keep the behavior consistent I made the count increment happen in the run loop code rather than the Dataset. (And the Callback is the analog of that decision.)

Luckily it turns out `StreamingFilesDataset` is more flexible than I had realized (thanks to @saeta for pointing this out!) and can stream the results of a `Dataset` running locally, effectively eliminating the need for the manual sharding/serialization/deserialization code that is currently in the TPU path. With that pattern, both GPU and TPU paths would start from the same `tf.data.Dataset.from_generator` and your intuition about not needing the Callback is correct. @tayo is currently looking at the TPU side of NCF.

So to make a long story short, that Callback can (and should) go away soon, but not quite yet. Hope that helps.

Reply:
Thanks for the explanation.

Reply:
@seemuch I am indeed working on the TPU path.  
@robieta Thanks for pointing out the StreamingFilesDataset simplification